modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -0,0 +1,627 @@
+            const depthType* depthRow = depth[y];
+            for (int x = 0; x < depth.cols; x += depthStride)
+            {
+                depthType z = depthRow[x] * depthFactor;
+                if (z <= 0)
+                    continue;
+
+                Point3f camPoint = reproj(Point3f((float)x, (float)y, z));
+                Point3f volPoint = cam2vol * camPoint;
+
+                //! Find accessed TSDF volume unit for valid 3D vertex
+                cv::Vec3i lower_bound = volume.volumeToVolumeUnitIdx(
+                    volPoint - cv::Point3f(volume.truncDist, volume.truncDist, volume.truncDist));
+                cv::Vec3i upper_bound = volume.volumeToVolumeUnitIdx(
+                    volPoint + cv::Point3f(volume.truncDist, volume.truncDist, volume.truncDist));
+                VolumeUnitIndexSet localAccessVolUnits;
+                for (int i = lower_bound[0]; i <= upper_bound[0]; i++)
+                    for (int j = lower_bound[1]; j <= upper_bound[1]; j++)
+                        for (int k = lower_bound[2]; k <= lower_bound[2]; k++)
+                        {
+                            const cv::Vec3i tsdf_idx = cv::Vec3i(i, j, k);
+                            if (!localAccessVolUnits.count(tsdf_idx))
+                            {
+                                localAccessVolUnits.emplace(tsdf_idx);
+                            }
+                        }
+                AutoLock al(mutex);
+                for (const auto& tsdf_idx : localAccessVolUnits)
+                {
+                    //! If the insert into the global set passes
+                    if (!volume.volumeUnits.count(tsdf_idx))
+                    {
+                        VolumeUnit volumeUnit;
+                        cv::Point3i volumeDims(volume.volumeUnitResolution,
+                                               volume.volumeUnitResolution,
+                                               volume.volumeUnitResolution);
+
+                        cv::Affine3f subvolumePose =
+                            volume.pose.translate(volume.volumeUnitIdxToVolume(tsdf_idx));
+                        volumeUnit.pVolume = cv::makePtr<TSDFVolumeCPU>(
+                            volume.voxelSize, subvolumePose, volume.raycastStepFactor,
+                            volume.truncDist, volume.maxWeight, volumeDims);
+                        //! This volume unit will definitely be required for current integration
+                        volumeUnit.index             = tsdf_idx;
+                        volumeUnit.isActive          = true;
+                        volume.volumeUnits[tsdf_idx] = volumeUnit;
+                    }
+                }
+            }
+        }
+    }
+
+    HashTSDFVolumeCPU& volume;
+    const Depth& depth;
+    const Intr::Reprojector reproj;
+    const cv::Affine3f cam2vol;
+    const float depthFactor;
+    const int depthStride;
+    mutable Mutex mutex;
+};
+
+struct VolumeUnitInFrustumInvoker : ParallelLoopBody
```

Comment:
consider using of lambda functions instead for all `ParallelLoopBody` stuff

Reply:
Do we prefer lambdas over functor objects just because of coding style or `ParallelLoopBody` misses some important features?

Reply:
I have changed a few small `parallel_for_` to lambdas.